core/src/test/java/org/apache/iceberg/TestMergeAppend.java

```diff
@@ -92,6 +94,74 @@ public void testEmptyTableAppend() {
         statuses(Status.ADDED, Status.ADDED));
   }
 
+  @Test
+  public void testEmptyTableAppendFilesWithDifferentSpecs() {
+    assertThat(listManifestFiles()).as("Table should start empty").isEmpty();
+
+    TableMetadata base = readMetadata();
+    assertThat(base.currentSnapshot()).as("Should not have a current snapshot").isNull();
+    assertThat(base.lastSequenceNumber()).as("Last sequence number should be 0").isEqualTo(0);
+
+    table.updateSpec().addField("id").commit();
+    PartitionSpec newSpec = table.spec();
+
+    assertThat(table.specs().size()).as("Table should have 2 specs").isEqualTo(2);
+
+    DataFile fileOriginalSpec =
+        DataFiles.builder(SPEC)
+            .withPath("/path/to/data-a.parquet")
+            .withPartitionPath("data_bucket=0")
+            .withFileSizeInBytes(10)
+            .withRecordCount(1)
+            .build();
+
+    DataFile fileNewSpec =
+        DataFiles.builder(newSpec)
+            .withPath("/path/to/data-b.parquet")
+            .withPartitionPath("data_bucket=0/id=0")
+            .withFileSizeInBytes(10)
+            .withRecordCount(1)
+            .build();
+
+    Snapshot committedSnapshot =
+        commit(
+            table, table.newAppend().appendFile(fileOriginalSpec).appendFile(fileNewSpec), branch);
+
+    assertThat(committedSnapshot).as("Should create a snapshot").isNotNull();
+    V1Assert.assertEquals(
+        "Last sequence number should be 0", 0, table.ops().current().lastSequenceNumber());
+    V2Assert.assertEquals(
+        "Last sequence number should be 1", 1, table.ops().current().lastSequenceNumber());
+
+    assertThat(committedSnapshot.allManifests(table.io()).size())
+        .as("Should create 2 manifests for initial write, 1 manifest per spec")
+        .isEqualTo(2);
+
+    long snapshotId = committedSnapshot.snapshotId();
+
+    validateManifest(
+        committedSnapshot.allManifests(table.io()).stream()
+            .filter(m -> Objects.equals(m.partitionSpecId(), SPEC.specId()))
+            .findAny()
+            .get(),
+        dataSeqs(1L),
+        fileSeqs(1L),
+        ids(snapshotId),
+        files(fileOriginalSpec),
+        statuses(Status.ADDED));
+
+    validateManifest(
+        committedSnapshot.allManifests(table.io()).stream()
+            .filter(m -> Objects.equals(m.partitionSpecId(), newSpec.specId()))
+            .findAny()
+            .get(),
+        dataSeqs(1L),
+        fileSeqs(1L),
+        ids(snapshotId),
+        files(fileNewSpec),
+        statuses(Status.ADDED));
```

Comment:
Nit: I think for this case you could just have a single validateManifest call and loop over the expected specs, something like:

```
for (Map.Entry<Long, DataFile> expectedFilesBySpec : ImmutableMap.of(newSpec, fileNewSpec, spec.specId(), fileOriginalSpec).entrySet()) {
   validateManifest(
        committedSnapshot.allManifests(table.io()).stream()
            .filter(m -> Objects.equals(m.partitionSpecId(), expectedFilesBySpec.getKey()))
            .findAny()
            .get(),
        dataSeqs(1L),
        fileSeqs(1L),
        ids(snapshotId),
        files(expectedFilesBySpec.getValue()),
        statuses(Status.ADDED));
}
```